### ISSUE TITLE
Adding other elements in mv-bar breaks right-hand alignment of mv-bar buttons

### ISSUE THREAD

**Issue by synicalsyntax** [mavoweb/mavo#693]
The current documentation example for customizing the Mavo toolbar allows users to add plaintext:
```html
<div mv-app mv-storage="local">
    <div class="mv-bar mv-ui" style="background: blue">
        YOLO
    </div>
    <div property="name">Lea</div>
</div>
```
![image](https://user-images.githubusercontent.com/15116870/106677643-12fe1880-65f4-11eb-9213-513b1ea06481.png)

However, in the event the toolbar text gets nested in an HTML element (`<p>`, `<span>`, etc.) the buttons lose their alignment to the right-hand side.

```html
<div class="mv-bar mv-ui">
    <div class="mv-bar mv-ui" style="background: blue">
        <span>YOLO</span>
    </div>
     <div property="name">Lea</div>
</div>
```

![image](https://user-images.githubusercontent.com/15116870/106677665-1f827100-65f4-11eb-9fdf-b9b6a45f7aaa.png)

If this isn't intended behavior, [changing](https://github.com/synicalsyntax/mavo/blob/master/src-css/_bar.scss#L70) `first-child` to `first-of-type` should allow the 2nd syntax to maintain the appearance of the original and thus resolve this issue.

**Comment by LeaVerou** [comment]
Hi Cynthia, nice to see you around! 👋🏼 

Ar first I was wondering "why would you nest two `.mv-bar` elements?", but it looks like the bug you're describing is present even with a plain HTML element. 

Here is a testcase: https://codepen.io/leaverou/pen/yLVYJOo?editors=1100
(you can easily make testcases by visiting https://play.mavo.io)

> Would be happy to open a PR for this if this is actually a bug and not intended behavior since it's a pretty simple change; [changing](https://github.com/synicalsyntax/mavo/blob/master/src-css/_bar.scss?rgh-link-date=2021-02-02T23%3A49%3A28Z#L70) `first-child` to `first-of-type` should allow the 2nd syntax to maintain the appearance of the original and thus resolve this issue.

Nice idea, and I think it would likely be an improvement in most cases. However, a serious issue with this proposed change is that we do have default buttons that are not `<button>` elements (such as `import` and `export`). Since `:first-of-type` matches on actual type (i.e. tag name), and not on selector, this means that when you have a `<span class="mv-button">` next to a `<button>`, both will get the margin because both are the first of their type. In the worst case, when you have all buttons, you'd get three buttons with the extra margin, as there are three used types (`<button>`, `<a>`, `<span>`). Essentially what we need here is to be able to say "the first element that matches `button, .mv-button`, but this is not possible to specify in CSS yet. In the future it will be, via [`:nth-child(1 of button, .mv-button)`](https://drafts.csswg.org/selectors-4/#nth-child-pseudo) but that is [not yet supported anywhere](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-child#browser_compatibility:~:text=of%20%3Cselector%3E%20syntax,-No). 😞 

Let's think how we can accomplish (something close to) what we want in a different way. What about this: The first child gets the `margin-left: auto`, *unless* it's not a `button, .mv-button`, in which case its next sibling gets it. It's not *exactly* what we want, because if we have two spans in the start, only the first will be left aligned, but it's close enough and is an improvement over the current code.

How could we serialize this as a CSS selector? Our first thought might be something like:

```css
:first-child:is(.mv-button, button),
:first-child:not(.mv-button, button) + * {
	margin-left: auto;
}
```

While this is as close as possible to the English sentence above, one downside is that [`:is()`](https://caniuse.com/css-matches-pseudo) and [selector lists in `:not()`](https://caniuse.com/css-not-sel-list) are relatively new and not very well supported. So we would need to "desugar" it manually:

```css
:first-child:.mv-button,
button:first-child,
:first-child:not(.mv-button):not(button) + * {
	margin-left: auto;
}
```

What do you think? It's entirely possible that I may have missed something, but if you think that makes sense, and it checks out when you try it, feel free to send that PR!

**Comment by synicalsyntax** [comment]
Hi @LeaVerou, thanks for the thorough explanation! I now understand why and agree that `:first-of-type` wouldn't be the right approach here.

However, I'm concerned about the implications of the third rule if we were to have multiple non-button/mv-button, native elements in the UI bar, as it would force any other HTML elements after the first to be aligned right as well. I've drawn up a [JSFiddle here](https://jsfiddle.net/egk2L09x/17/) illustrating this. I don't think it's best to assume the user will only be providing one non-button element in the UI bar, so I think it'd be best to not declare a `:first-child` to detect that element and strictly use the button/mv-button sibling selectors right after it. Let me know what you think!

**Comment by joyously** [comment]
Were you going to change it to `margin-right` for RTL languages?

**Comment by LeaVerou** [comment]
@synicalsyntax If my understanding is correct, you're proposing to use this selector instead:

```css
:not(.mv-button):not(button) + .mv-button,
:not(.mv-button):not(button) + button
```

While that works well in the case in the JSFiddle, it means there could be more than one `margin-left: auto` inserted, which we *definitely* don't want. E.g. consider this: https://jsfiddle.net/jh3p06sf/

@joyously We could, it's just a matter of using logical properties instead of physical. However, this is orthogonal, could you please open a new issue about it?

**Comment by synicalsyntax** [comment]
@LeaVerou  Yep in regards to the selector. If we want to account for all these scenarios, I'm not sure if there's a pure CSS way of going about this. Would it be reasonable to wrap any Mavo-injected buttons in a parent div on the JS side and apply the `margin-left: auto` property to that?

**Comment by LeaVerou** [comment]
But Mavo injected buttons are no different than author-injected buttons in that regard. There is no reason to have Mavo injected buttons on the right and author-injected buttons on the left.

We *could* have JS to deal with this, but that increases maintenance cost, for little benefit, since this is something that authors may wish to customize anyway, we're just providing sensible defaults, and there's nothing inherently wrong with a button being on the left or a div being on the right. I think as long as we avoid applying the margin on multiple elements (which looks terrible), it's not a huge deal if it's not applied in the optimal place. So the question is, what CSS would give us "good enough" results, that are an improvement over the current behavior in every case?

**Comment by synicalsyntax** [comment]
So from my understanding, our current goal is to have the default be the first non-button/mv-button element aligned to the left, and everything else coming after it on the right, regardless if the following elements are buttons or non-buttons (`<p>`/`<span>`). It would be up to the author to erase the `margin-left: auto` on the desired element and reset it on its following element: https://jsfiddle.net/ry592c8h/2/

 If that's the case, the CSS you originally suggested would work. Please correct me if there's any misunderstandings!
